url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+			while (callFrame.isFinished()) {
+				if (stackpointer == callFrame.basepointer) {
+					push(XValueNull.NULL);
+				} else if (stackpointer > callFrame.basepointer + 1) {
+					XValue ret = pop();
+					pop(stackpointer - callFrame.basepointer);
+					push(ret);
+				}
+				callFrame = callFrame.getParent();
+				if (callFrame == null)
+					return instrs;
+			}
+			if (exception != null) {
+				XCatchHandler catchHandler = callFrame.getCatchHandler();
+				while (catchHandler == null) {
+					callFrame = callFrame.getParent();
+					if (callFrame == null)
+						return instrs;
+					catchHandler = callFrame.getCatchHandler();
+				}
+				int stackpointer = catchHandler.getStackPointer();
+				pop(this.stackpointer - stackpointer);
+				push(exception);
+				exception = null;
+				callFrame.jumpTo(catchHandler.getInstructionPointer());
+			}
+			try {
+				exec();
+			} catch (XRuntimeException e) {
+				XValue value = rt.getModule("sys");
+				XValue type = value.getRaw(rt, e.getType());
+				XValue exc = rt.alloc(type);
+				XValue message = rt.alloc(e.getMessage());
+				XValue stackTrace = XUtils.convertStackTrace(rt, getStackTrace(false));
+				exc.setRaw(rt, "_message", message);
+				exc.setRaw(rt, "_cause", XValueNull.NULL);
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			} catch (Throwable e) {
+				StringWriter sw = new StringWriter();
+				e.printStackTrace(new PrintWriter(sw));
+				String message = sw.toString();
+				XValue value = rt.getModule("sys");
+				XValue type = value.getRaw(rt, "NativeError");
+				XValue exc = rt.alloc(type);
+				XValue m = rt.alloc(message);
+				XValue stackTrace = XUtils.convertStackTrace(rt, getStackTrace(false));
+				exc.setRaw(rt, "_message", m);
+				exc.setRaw(rt, "_cause", XValueNull.NULL);
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			}
+		}
+		return instrs;
+	}
+
+	private void exec() {
+		int inst = callFrame.readUByte();
+		XOpcode opcode = XOpcode.values()[inst];
+		XValue v1, v2, v3, v4;
+		int i1;
+		String s;
+		if (XFlags.DEBUG) {
+			System.out.println("exec:" + callFrame.getMethodName() + ":" + opcode + ":" + Arrays.toString(Arrays.copyOf(stack, stackpointer)));
+		}
+		switch (opcode) {
+		case ADD:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isNumber() && v2.isNumber()) {
+				if (v1.isFloat() || v2.isFloat()) {
+					v3 = new XValueFloat(v2.getFloat() + v1.getFloat());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() + v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator("+", "__add__", "__add2__", v2, v1);
+			}
+			break;
+		case AND:
+			v1 = pop();
+			v2 = pop();
+			if ((v1.isInt() || v1.isBool()) && (v2.isInt() || v2.isBool())) {
+				if (v1.isBool() && v2.isBool()) {
+					v3 = XValueBool.valueOf(v2.getBool() && v1.getBool());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() & v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator("&", "__and__", "__and2__", v2, v1);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 575:584, 1445:1454, 1787:1796.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.